README
Copyright (C) 2003-2006 Jeffrey I Cohen.  All rights reserved, worldwide.

NAME

  Genezzo - an extensible database system

DESCRIPTION

  The Genezzo modules implement a hierarchy of persistent
  hashes using a fixed amount of memory and disk.  This system is
  designed to be easily configured and extended with custom functions,
  persistent storage representations, and novel data access methods.
  In its current incarnation it supports a limited subset of SQL with
  a command line tool [gendba] and a partial DBI-style interface.

REQUIREMENTS

 Perl 5.6    

INSTALLATION

  If you can read this, then you figured out how to uncompress and
  untar the archive.  

  perl Makefile.PL  # constructs the makefile
  make              
  make test         
  make install      # (as superuser [optional])


  If you install Genezzo, you should be able to run the line-mode tool
  directly by typing "gendba.pl" at the command line.

  You also should be able to run scripts from the directory containing
  this file, something like:

  perl -Iblib/lib lib/Genezzo/gendba.pl


USAGE

  line mode available via (/usr/bin) gendba.pl.  
  Try:

    gendba.pl -help
    gendba.pl -man 

  or:

    perl -Iblib/lib lib/Genezzo/gendba.pl -help
    perl -Iblib/lib lib/Genezzo/gendba.pl -man

  for more information.

  The GenDBI module supports a subset of DBI-style calls.  See
  dbi_example.pl for some sample usage.

  Try:

    perl -Iblib/lib demo/dbi_example.pl # runs an example
    perldoc Genezzo::GenDBI

  for more information.

  The simplest way to create an instance of a Genezzo database is to use:

    gendba.pl -init

  This command will create a new database and login to the
  command-line.  

SQL SUPPORT -- see LIMITATIONS

  EXPRESSIONS: The Genezzo parser can handle expressions composed of
               column names, numeric or quoted string constants,
               logical constants like TRUE, FALSE and NULL,
               concatenation, scalar functions, and basic arithmetic
               like '+', '-', '*', '/'.  However, logical expressions
               using AND, OR, and NOT are only legal in the WHERE clause.

  INSERT: 
    INSERT INTO tablename VALUES (expr [, expr ...]);
    INSERT INTO tablename (colname [, colname ...]) VALUES (expr [, expr ...]);
    INSERT INTO tablename SELECT ... FROM ...

  UPDATE:
    UPDATE tablename SET colname = expr [, colname = expr ...] [WHERE ...]

  DELETE:
    DELETE FROM tablename [WHERE ...]

  SELECT:  The Genezzo parser can parse complex SELECT statements,
           but the execution engine can only handle simple SELECTS
           and basic joins.

    SELECT * FROM tablename [WHERE ...]
    SELECT COUNT(*) FROM tablename 
    SELECT ECOUNT(*) FROM tablename
    SELECT * FROM t1, t2 WHERE t1.col1 = t2.col1    

    SELECT expr [[AS] column_alias] [, expr [[AS] column_alias]]
      FROM tablename
      [WHERE ...]

  CREATE TABLE: 
    CREATE TABLE tablename (colname coltype [, colname coltype ...]);

  CREATE INDEX:
    CREATE INDEX iname on tablename (colname [, colname ...]);

  ALTER TABLE ADD CONSTRAINT: supports primary and unique key and
    check constraints.  NOT NULL and foreign keys are not currently
    supported, though you can implement NOT NULL as 
    CHECK (colname IS NOT NULL).  The CHECK expression is equivalent
    to a WHERE clause.  A table can have multiple CHECK constraints,
    but currently only one key constraint.

    ALTER TABLE tablename 
          ADD [CONSTRAINT constraint_name] 
              PRIMARY KEY (colname [, colname ...]);
    ALTER TABLE tablename 
          ADD [CONSTRAINT constraint_name] 
              UNIQUE (colname [, colname ...]);
    ALTER TABLE tablename 
          ADD [CONSTRAINT constraint_name] 
              CHECK (conditional_expression);

  ADDFILE: Not really a SQL command, but an extension to grow the 
           datafiles.  See "addfile help".

  DROP: drop a table or index.
  COMMIT: write changes to disk.
  SYNC: write changes to disk like "commit", but don't end a transaction.
  ROLLBACK: discard all uncommitted changes.

  DESCRIBE: 
    DESCRIBE tablename

  EXPLAIN PLAN: dump parsing and plan information for any SQL command,
                including CREATE and ALTER TABLE.  Does not require a
                "plan table".  Can perform partial explanations for 
                invalid statements, such as missing tables/columns, etc.
                Example - "explain plan for select * from t1;"
                or even complex sql like:
                Explain plan for
                  select * from t1 
                           where c1 in (select avg(c2) from t2)
                           having avg(c3) > ((select min(c4)
                                             from (select c5 from t3
                                                   union    
                                                   select c6 from t4)
                                             ) + 5) * 7 / min(c7)     ;

  STARTUP/SHUTDOWN: 

LIMITATIONS

  General: No locking/multiuser support, bind variables,          
  incomplete DBI support.

  SELECT: Simple queries only -- no ORDER BY, GROUP BY, 
          HAVING, subqueries, set functions (UNION, INTERSECT).
          Only basic FROM clause joins: FROM t1, t2, ... are
          supported.
          No aggregation functions other than COUNT(*), ECOUNT(*).
          COUNT and ECOUNT do not accept a WHERE clause.
  
  ALTER TABLE: no foreign key, NULL/NOT NULL constraints.

  CREATE TABLE: Only numeric and character datatypes are accepted.
                Datatype length and precision are not enforced.
                Numeric columns will accept arbitrary character input,
                but will attempt to use numeric (versus character) 
                comparisons.  No table or column constraints.
                No CREATE TABLE AS SELECT.

EXAMPLES
        
  Some very basic SQL create/drop/describe table, select, insert, update
  and delete syntax is supported, and like standard SQL, table and
  column names are case-insensitive unless quoted.  More complex SQL
  parses, but is ignored.  While many scalar SQL functions are available, 
  the only supported aggregation functions are count(*) and ecount(*),
  a non-blocking count estimation function.  The database also
  supports commit to force changes to disk, and a crude rollback. 
  NOTE: Data definition (such as create table or ct) must be manually
  committed to keep the database in a consistent state.  Uncommitted
  inserts and updates will only be block-consistent -- there is no
  guarantee that the data will get flushed to disk, and no
  guarantee whether the changes will or will not take effect.

    rem  Some simple SELECTs
    select * from _col1;
    select rid, rownum, tname, colname from _col1;
    select count(*) from _col1;
    select ecount(*) from _col1;

    rem  SELECTs with WHERE, perl and SQL style.
    select * from _tab1 where tname =~ m/col/;
    select * from _tab1 where tid < 5;
    select * from _tab1 where numcols > 3 AND numcols < 6;

    rem basic joins
    select * from _tab1, _col1  where _tab1.tid = _col1.tid 
     and _tab1.tname =~ m/col/;

    rem  Column aliases and Expressions
    select tid as Table_ID, tname Name, (tid+1)/2 from _tab1;

    rem  Basic INSERT
    insert into test1 values ('a','b','c','d');
    insert into test1(col2, col1) values ('a','b','c','d');

    rem CREATE TABLE and INSERT...SELECT
    create table test2(col1 char, col2 char);
    insert into test2 (col1) select col1 from test1;

    rem  DELETE with WHERE
    delete from test1 where col1 < 'bravo' and col2 > 5;

    rem  UPDATE with WHERE (no subqueries supported)
    update test2 set col2 = 'foo' where col2 is null;

    commit

  By default, the database will autostart -- the gendba.pl command will
  connect to an active, updateable instance.  The shutdown command will
  reset the database to a read-only mode that only provides read access
  to certain dictionary tables.  The startup command will restart the
  database.

  The Genezzo parser also supports the Feeble query language, as
  described by gendba.pl -man.  Briefly, the commands are ct (create
  table), dt (drop table), i (insert), u (update), d (delete), s
  (select).  The following example creates a table EMP with two
  columns, ENAME (as character) and ID (as number), inserts five rows,
  and counts them.

    ct EMP ENAME=c ID=n
    i EMP bob 1 orville 2 wilbur 3 harry 4 luke 5
    s EMP count
    s EMP ecount
    commit

  While the SQL update and delete commands support a WHERE clause, the
  Feeble "u" and "d" commands only work via rowid.  For example:

    gendba 10> s EMP rid rownum *
    rid     rownum  ename   id
    ___     ______  _____   __

    1/33/1  1       bob     1
    1/33/2  2       orville 2
    1/33/3  3       wilbur  3
    1/33/4  4       harry   4
    1/33/5  5       luke    5

    5 rows selected.

    gendba 11> d EMP 1/33/1
    deleted 1 rows from table emp.

    gendba 12> u EMP 1/33/5 margo 66
    updated 1 rows in table emp.

    gendba 13> !10
    s EMP rid rownum *
    rid     rownum  ename   id
    ___     ______  _____   __

    1/33/2  1       orville 2
    1/33/3  2       wilbur  3
    1/33/4  3       harry   4
    1/33/5  4       margo   66

    4 rows selected.

  The default database is stored in a fixed-size 160K file.  You can
  change the default filesize and the blocksize at database
  initialization time.  For example:

    gendba.pl -init -define blocksize=8k -define dbsize=10M

  to create a database with a 10 megabyte datafile using 8K blocks.
  By default, this file is a fixed size, but it can be set to grow
  using the "increase_by" settings added in version 0.18.  Also, you can
  add additional files to your database using the "addfile" command.  By
  default, each file is double the size of the previous.  Type "addfile
  help" for more options.


NEW FEATURES - Indexed Tables

  Version 0.13 introduces an indexed-table type.  An indexed table
  treats the first column in the table definition as a unique, not-null
  primary key.  Note that an "Indexed Table" is distinct from an
  ordinary table with a separate index, where you can assign a separate
  index object to a particular column or group of columns.

    gendba 9> ct EMP index id=n name=c
    Create Table : emp 
    with unique index option
    tablename : emp
    column id : n (primary key)
    column name : c
    table emp created 

    gendba 10> i EMP 1 bob 2 orville 3 wilbur 4 harry 5 luke
    inserted 5 rows into table emp.

    gendba 11> s EMP rid *
    rid     id      name
    ___     __      ____

    1       1       bob
    2       2       orville
    3       3       wilbur
    4       4       harry
    5       5       luke

    5 rows selected.

    gendba 12> i EMP 1 bob 2 orville 3 wilbur 4 harry 5 luke
    whisper: duplicate key 1
    Failed to insert row 1 in table emp at lib/Genezzo/gendba.pl line 230
    inserted 0 rows into table emp.

    gendba 13> select tid, tname, otype from _tab1 where tname = 'emp'
    tid     tname   otype
    ___     _____   _____

    14      emp     IDXTAB

    1 row selected.


  Notice that the first column is marked as a primary key when the
  table is created, and the rids for indexed tables are the primary key,
  not a physical block address.  

  Indexed tables have an object type of IDXTAB in the data dictionary
  -- conventional tables are marked as TABLE.

  WHERE clause processing has been enhanced to use equality predicates
  (e.g. "SELECT * from EMP where id = 2") to drive index scans.  Any
  series of "AND"ed predicates is a candidate for an efficient index
  search, but the use of an OR currently requires a full table scan.

  CAVEATS:  The index performance is pretty poor, for a couple reasons.
  It should improve a lot.  

NEW FEATURES - Indexes

  Users can now define their own indexes.  The dictionary
  automatically maintains indexes on its own tables.  In order to create
  an index called EMP_I1 on the ID column of table EMP, the command is:
  
      Create Index EMP_I1 on EMP (ID)

  You can have arbitrary combinations of numeric and character keys.
  RESTRICTIONS: only one index per table currently.  Index keys must
  be less than one-half of the db blocksize.  

  In version 0.15, the query engine will use the index to drive the
  row fetch for a SELECT if the WHERE clause specifies an equality 
  predicate on the primary key.  For example:

      Select * from EMP where ID = 5
 
  uses an index.  In general, additional predicates may accompany the
  equality expression as long as they are ANDed:

      Select * from EMP where ID = 5 and ENAME =~ m/BOB/

  However, the presence of ORs causes the plan to revert to a full
  table scan:

      Select * from EMP where ID = 5 or ID = 6

  All user-defined indexes are non-unique.  Unique indexes should be
  created as unique constraints (fixed in version 0.22) using 
  ALTER TABLE.

NEW FEATURES - automatic tablespace and file growth

  In version 0.18, datafiles are no longer a fixed size.  The column
  "increase_by" in table _tsfiles controls the behavior: if this column
  is blank, null, or zero, the datafile size is fixed.  If
  "increase_by" is a non-zero number, the datafile will attempt to grow by 
  that many bytes if it runs out of free extents.  If "increase_by" is a
  percentage like "50%", the datafile is extended by the current size
  multiplied by this percentage.  The AddFile command now takes
  "increase_by" as an optional argument.  

  Also in version 0.18, each tablespace can be configured to
  automatically create new datafiles when the current datafile runs out
  of free extents.  The column "addfile" in table _tspace controls the
  behavior: if the column is blank or null, then the tablespace must be
  extended manually.  If the "addfile" column contains a valid argument 
  list for the AddFile command, then this command will get invoked when 
  all of the current datafiles are full.  Note that the datafiles can be 
  of a fixed size, or use the new "increase_by" parameter to grow to some
  maximum size (currently 2G).

  Prior to version 0.69, the simplest way to alter tablespace settings
  was the "bigdb.sql" script.  This script is supplanted in version 0.69
  by the alter_ts() utility function.  The command:
    select alter_ts() from dual;
  will, by default, upgrade the SYSTEM tablespace to use a base file
  size of 10M which automatically grows in 50% increments.  For more 
  information, in gendba.pl line-mode please type:
    help area=utility_functions alter_ts
  
NEW FEATURES - ALTER TABLE ADD CONSTRAINT

  Version 0.20 supports ALTER TABLE ADD CONSTRAINT <name> CHECK (<check-text>).

  The check-text is any basic where clause predicate.  Example:

    gendba 3> create table testcon (col1 c, col2 n);
    table testcon created 
    
    gendba 4> alter table testcon add constraint cn1 
              check ((col2 > 5) AND col1 =~ m/(a|b|c)/)
    Added constraint cn1 to table testcon
    
    gendba 5> insert into testcon values ('a', 7);
    inserted 1 row into table testcon.
    
    gendba 6> insert into testcon values ('b', 8);
    inserted 1 row into table testcon.
    
    gendba 7> insert into testcon values ('d', 8);
    WARNING: violated constraint cn1
    must satisfy (col2 > 5) AND col1 =~ m/(a|b|c)/
    WARNING: Failed to insert row 1 in table testcon
    inserted 0 rows into table testcon.
    
    gendba 8> insert into testcon values ('a', 1);
    WARNING: violated constraint cn1
    must satisfy (col2 > 5) AND col1 =~ m/(a|b|c)/
    WARNING: Failed to insert row 1 in table testcon
    inserted 0 rows into table testcon.


  Version 0.22 supports 
    ALTER TABLE ADD CONSTRAINT <name> UNIQUE (<column-list>)
    and
    ALTER TABLE ADD CONSTRAINT <name> PRIMARY KEY (<column-list>

  Version 0.22 only supports a single constraint or single index per
  table, but version 0.23 supports multiple constraints and indexes
  on a table.
  
  Need to extend to fully support NULL/NOT NULL, FOREIGN KEY, plus
  ENABLE/DISABLE.

NEW FEATURES - System/User Extensibility

  Version 0.25 of Genezzo supports database extensibility.  See
  Genezzo::Havok for more details.  The script "havok.sql" (now
  deprecated) can be used to add extensibility tables to your database.
  In version 0.58 a new function HavokUse supercedes the havok.sql
  script, and the table user_extend is supplanted in version 0.59 by
  Havok::UserFunctions.

NEW FEATURES - Improved error reporting, handling

  Version 0.31 error reporting has been significantly reworked to be
  more compatible with DBI.  In particular, Genezzo::GenDBI->connect now

Changes
Revision history for Perl extension Genezzo

0.71 Wed Jul  4 00:03:46 2007
        - change output file spooling (drop tee dependency)
        - fixes for Havok, SysHook MakeSQL (havok.sql/syshook.sql)
        - SMExtent: self-managed extents (disabled)
        - help: slight regex fixups
        - SysHook: add support for dynamic load of new hooks into 
          running database instance (without restart)
        - line-mode history fixes

0.70 Mon Jan 15 01:02:45 2007
        - improved help
        - substr fix
        - fix TODO generation
        - history formatting

0.69 Sun Dec  3 01:39:32 2006
        - add_user_function: add support to dynamically load new
          functions into running database instance (without restart).
        - alter_ts: function to alter tablespaces

0.68 Sat Oct 28 23:54:03 2006
        - Like, In: new sql comparison functions
        - SQL scalar functions: fixes for initcap, ltrim, rtrim

0.67 Sun Oct 22 23:26:26 2006
        - Added SQLTest for simple, log-based scripted testing.
        - SQL scalar functions: fixes for crypt, pack, sprintf, rindex, 
          int, log.  
        - Added lpad, ltrim, rpad, rtrim, translate, replace, soundex.
        - Added mod, power, round, sign, trunc, logN.

0.66 Fri Oct 20 11:49:27 2006
        - fix sprintf SQL function
        - move dbi_example to demo directory
        - more pod fixes to improve kwalitee

0.65 Thu Oct 19 02:26:57 2006
        - Parse/Plan: position information for identifiers
        - better support for rownum filters
        - fixes for absolute filenames for datafiles
        - pod fixes to improve kwalitee

0.64 Mon Aug 21 14:10:54 2006
        - enhancements to debug utility functions

0.63 Fri Aug 11 00:56:28 2006
        - fix Feeble case-sensitivity (Andrew, thanks again)
        - several fixes for quoted identifiers with embedded spaces

0.62 Sat Aug  5 15:59:52 2006
        - fix dbi_example.pl (thanks, Andrew)
        - debug utility functions

0.61 Fri Jul  7 00:26:15 2006
        - RDBlock hooks
        - RSTab update improvements
        - preliminary SMExtent/SMHook packaging

0.60 Fri May 12 22:53:12 2006
        - genexp enhancements
        - quurl/unquurl: new sql functions for URL-style quoting

0.59 Sun May  7 00:20:50 2006
        - top-level Genezzo.pm
        - Havok: rename UserExtend to UserFunctions.
        - Havok: add scalar SQL functions
        - fix ECOUNT parsing in select 
        - rectify database initialization chattiness

0.58 Mon Apr  3 00:48:51 2006
        - Havok: new HavokUse function to load Havok modules.  
        - INSERT SELECT: fix null padding for insufficient columns
        - add DUAL table 

0.57 Fri Mar 10 22:49:16 2006
        - extent management in table segments - preliminary work.
        - restrict buffer cache to 100 open files.
        - require a minimum of 4 blocks per file.

0.56 Mon Jan  2 02:41:05 2006
        - Fix "uninitialized value" for invalid column names in insert 
          (thanks, Eric)
        - Added gendba and genexp "-version" flag.
        - Block checksum utility functions.
        - DBI extension: parse_tree_prepare method - prepare a parse_tree 
                         (versus prepare a sql statement).
        - fix bigdb.sql to increase buffer cache size.

0.55 Mon Dec 12 01:17:52 2005
        - Fix "Drop Table" case-sensitivity (thanks, Andrew)
        - Util::USE_FSYNC: check for missing fsync on Win32 and autoflush
          instead

0.54 Wed Nov 30 23:58:53 2005
        - Documentation fixes
        - quiet debug info in QueryRewrite (thanks, Eric)

0.53 Sun Nov 27 20:32:23 2005
        - Documentation fixes

0.52 Fri Nov 25 18:15:16 2005
        - fix unary expression evaluation
        - use TestSetup for regress
        - extend type checking for FROM clause subqueries, set operations
        - basic join support

0.51 Sun Sep 18 00:44:17 2005
        - dictionary recursive sql support
        - CREATE TABLESPACE, add TABLESPACE support to CREATE TABLE/INDEX

0.50 Wed Sep  7 01:36:18 2005
        - add correspondence between BufCaElt and RDBlock via the
          MailBag
        - add hooks for _BCE_dirtyhook, _init_filewriteblock

0.49 Sun Aug 28 22:24:16 2005
        - Util::FileSetHeaderInfo - method to add defs to existing datafile

0.48 Fri Aug 26 23:33:21 2005
        - Consolidate all dictionary defs/prefs, see "dump prefs"

0.47 Thu Aug 25 02:03:21 2005
        - Object-Oriented SysHook
        - arbitrary user preferences (via -define) and fileheader info
          (via -fileheader_define).
        - align fileheader on blocksize boundary

0.46 Sun Aug  7 20:10:17 2005
        - more efficient update of rows that exceed small row threshold.
        - add sync (flush changes to disk without committing a transaction).

0.45 Sat Jul 23 22:26:29 2005
        - More fixes for Raw IO:
        - Util: fix FileGetHeaderInfo for Raw IO - increase buffer size
          to handle large, aligned file headers.
        - Include file header size in space calculations for file size
          and number of blocks -- don't extend past specified file size.
        - add description field to _pref1 dictionary table

0.44 Sun Jul 17 00:34:51 2005
        - Eric's fixes for my merge of Raw IO
        - Fix FSF address everywhere

0.43 Fri Jul  8 02:34:17 2005
        - Enforce case-insensitivity in table, column names.
        - Foundation work for multiple tablespaces.
        - Update address of FSF (spotted by the 
          sharp eyes of Darren Duncan!)
        - Merge Eric Rollins' support for Raw IO.
        - Fix Syshook documentation (thanks again, Eric!)

0.42 Fri Jun 17 01:31:05 2005
        - fix boolean NOT
        - Chain GZERR routines for Havok, UserExtend and other modules
        - Add genexp.pl to install
        - New parser, execution for UPDATE, CREATE 
        - retire Feeble parser

0.41 Thu May 26 00:42:38 2005
        - allow TRUE, FALSE, and NULL in SELECT lists.

lib/Genezzo/XEval.pm
#!/usr/bin/perl
#
# $Header: /Users/claude/fuzz/lib/Genezzo/RCS/XEval.pm,v 7.4 2005/11/26 01:57:30 claude Exp claude $
#
# copyright (c) 2005 Jeffrey I Cohen, all rights reserved, worldwide
#
#
package Genezzo::XEval;
use Genezzo::Util;

use Genezzo::XEval::Prepare;
use Genezzo::XEval::SQLAlter;

use strict;
use warnings;
use warnings::register;

use Carp;

our $VERSION;

BEGIN {
    $VERSION = do { my @r = (q$Revision: 7.4 $ =~ /\d+/g); sprintf "%d."."%02d" x $#r, @r }; # must be all one line, for MakeMaker

}

our $GZERR = sub {
    my %args = (@_);

    return 
        unless (exists($args{msg}));

    if (exists($args{self}))
    {
        my $self = $args{self};
        if (defined($self) && exists($self->{GZERR}))
        {
            my $err_cb = $self->{GZERR};
            return &$err_cb(%args);
        }
    }

    my $warn = 1;
    if (exists($args{severity}))
    {
        my $sev = uc($args{severity});
        $sev = 'WARNING'
            if ($sev =~ m/warn/i);

        # don't print 'INFO' prefix
        if ($args{severity} =~ m/info/i)
        {
            printf ("%s: ", $sev);
            $warn = 0;
        }

    }
    carp $args{msg}
      if (warnings::enabled() && $warn);
    
};

sub _init
{
    my $self = shift;
    my %args = (@_);

    return 0
        unless (defined($args{plan}));

    $self->{plan} = $args{plan};
    $self->{prepare} = Genezzo::XEval::Prepare->new();
    
    return 1;
}

sub new 
{
    my $invocant = shift;
    my $class = ref($invocant) || $invocant ; 
    my $self = { };
    
    my %args = (@_);

    if ((exists($args{GZERR}))
        && (defined($args{GZERR}))
        && (length($args{GZERR})))
    {
        # NOTE: don't supply our GZERR here - will get
        # recursive failure...
        $self->{GZERR} = $args{GZERR};
        my $err_cb     = $self->{GZERR};
        # capture all standard error messages
        $Genezzo::Util::UTIL_EPRINT = 
            sub {
                &$err_cb(self     => $self,
                         severity => 'error',
                         msg      => @_); };
        
        $Genezzo::Util::WHISPER_PRINT = 
            sub {
                &$err_cb(self     => $self,
#                         severity => 'error',
                         msg      => @_); };
    }

    return undef
        unless (_init($self, %args));

    return bless $self, $class;

} # end new

# get or set the dictionary object
sub Dict
{
    my $self = shift;

    if (scalar(@_))
    {
        $self->{dictobj} = shift;
    }
    return $self->{dictobj};    
}

sub Prepare
{
    whoami;

    my $self = shift;

    my %required = (
                    plan => "no plan!"
                    );

    my %args = ( # %optional,
                @_);

    my ($msg, %earg);

    return undef
        unless (Validate(\%args, \%required));

    my $alg = $args{plan};

    return ($self->{prepare}->Prepare(plan => $alg,
                                      dict => $self->Dict()));

}

sub SQLAlter
{
    whoami;

    my $self = shift;

    my %required = (
                    plan => "no plan!"
                    );

    my %args = ( # %optional,
                @_);

    my ($msg, %earg);

    return undef
        unless (Validate(\%args, \%required));

    my $alg = $args{plan};

    return 0
        unless (exists($alg->{sql_alter}));

    if (exists($alg->{sql_alter}->{add_table_cons}))
    {
        my $add_tab_cons = $alg->{sql_alter}->{add_table_cons};

        my $tablename = $alg->{sql_alter}->{tc_table_fullname};
        
        my $cons_name;

        if (scalar(@{$add_tab_cons->{name}}))
        {
           $cons_name = 
             $add_tab_cons->{name}->[0]->[0]->{bareword};
        }

        greet $tablename, $cons_name;

        my %nargs = (
                     tname   => $tablename,
                     dbh_ctx => $args{dbh_ctx}
                     );

        if (defined($cons_name))
        {
            $nargs{cons_name} = $cons_name;
        }
        
        if (exists($add_tab_cons->{constraint}) &&
            exists($add_tab_cons->{constraint}->{cons_type}) &&
            ($add_tab_cons->{constraint}->{cons_type} 
             =~ m/check|primary|unique/i))
        {
            $nargs{cons_type} = $add_tab_cons->{constraint}->{cons_type};
        }
        else
        {
            $msg = "unknown constraint\n";
            $msg .= Data::Dumper->Dump( [%nargs]);
            %earg = (self => $self, msg => $msg,
                     severity => 'warn');
            
            &$GZERR(%earg)
                if (defined($GZERR));
            
            return 0;
        }

        if ($add_tab_cons->{constraint}->{cons_type} 
            =~ m/primary|unique/i)
        {

            # XXX XXX XXX: need to move these checks to typecheck
            my %dupi;

            # check the column list for duplicates
            for my $col (@{$add_tab_cons->{constraint}->{tc_column_list}})
            {
                if (exists($dupi{$col}))
                {
                    $msg = 'Duplicate column (' . $col . ') ';
                    $msg .= 'in constraint declaration';
                    %earg = (self => $self, msg => $msg,
                             severity => 'warn');
                    
                    &$GZERR(%earg)
                        if (defined($GZERR));
            
                    return 0;

                }
                else
                {
                    $dupi{$col} = 1;
                }
            }


            $nargs{cols} = $add_tab_cons->{constraint}->{tc_column_list};
        }
        elsif ($add_tab_cons->{constraint}->{cons_type} 
               =~ m/check/i)
        {
            my $where_clause =
                $add_tab_cons->{constraint}->{operands}->{sc_txt};
            $nargs{where_clause} = $where_clause;

            # needs to be an array to match WHERE clause
            my $where_arr = [
                $add_tab_cons->{constraint}->{operands}
                             ];

            my $where_filter = 
                $self->{plan}->SQLWhere2(tablename => $tablename,
                                         where     => $where_arr
                                         );

            unless (defined($where_filter))
            {
                $msg = "invalid where clause";
                %earg = (self => $self, msg => $msg,
                            severity => 'warn');
                    
                &$GZERR(%earg)
                    if (defined($GZERR));
                return 0;
            }
            $nargs{where_filter} = $where_filter->{filter_text};
        }

        greet %nargs;

        my ($stat, $new_consname, $new_iname) = 
            $self->{dictobj}->DictTableAddConstraint(%nargs);
        
        my $severity;
        if ($stat)
        {
            $cons_name = $new_consname
                unless (defined($cons_name));
            $msg = "Added constraint $cons_name" .
                " to table $tablename\n";
            $severity = 'info';
        }
        else
        {
            $msg = "Failed to add constraint\n";
            $severity = 'warn';
        }
        %earg = (self => $self, msg => $msg,
                 severity => $severity);
        
        &$GZERR(%earg)
            if (defined($GZERR));
        
        return $stat;
        
    } # end alter table constraint
    
    $msg = "cannot execute ALTER command";
    %earg = (self => $self, msg => $msg,
             severity => 'warn');
                    
    &$GZERR(%earg)
        if (defined($GZERR));

    return 0;
} # end SQLAlter

sub SQLInsert
{
    whoami;

    my $self = shift;

    my %required = (
                    plan => "no plan!"
                    );

    my %args = ( # %optional,
                @_);

    my ($msg, %earg);

    return undef
        unless (Validate(\%args, \%required));

    my $alg = $args{plan};

    unless (exists($alg->{sql_insert}) &&
            exists($alg->{sql_insert}->[1]->{insert_values}))
    {
        $msg = "cannot execute INSERT command";
        %earg = (self => $self, msg => $msg,
                 severity => 'warn');
        
        &$GZERR(%earg)
            if (defined($GZERR));
        
        return undef;
    }
    
    if (exists($alg->{sql_insert}->[0]->{insert_tabinfo}->{tc_column_list}))
    {
        my $tabinfo = $alg->{sql_insert}->[0]->{insert_tabinfo};


        # XXX XXX XXX: need to move these checks to typecheck
        my %dupi;

        # check the column list for duplicates
        for my $col (@{$tabinfo->{tc_column_list}})
        {
            if (exists($dupi{$col}))
            {
                $msg = 'Duplicate column (' . $col . ') ';
                $msg .= 'in INSERT';
                %earg = (self => $self, msg => $msg,
                         severity => 'warn');
                
                &$GZERR(%earg)
                    if (defined($GZERR));
            
                return undef;

            }
            else
            {
                unless (exists($tabinfo->{tc_table_colhsh}->{$col}))
                {
                    $msg =  'No such column ('. $col . ') ';
                    $msg .= 'in table ' . $tabinfo->{tc_table_fullname};
                    $msg .= ' for INSERT';
                    %earg = (self => $self, msg => $msg,
                             severity => 'warn');
                    
                    &$GZERR(%earg)
                        if (defined($GZERR));
            
                    return undef;

                }

                $dupi{$col} = 1;
            }
        }
    }

    # standard INSERT into ... VALUES ...
    if (ref($alg->{sql_insert}->[1]->{insert_values}) eq 'ARRAY')
    {
        my $tabinfo = $alg->{sql_insert}->[0]->{insert_tabinfo};

        use Genezzo::Row::RSExpr;
        use Genezzo::Row::RSDual;

        my @sel_list;
        
        for my $val (@{$alg->{sql_insert}->[1]->{insert_values}})
        {
            push @sel_list, { value_expression => $val};
        }
        greet @sel_list;
        
        my %tempo;
        my $rsd_tv = tie %tempo, 'Genezzo::Row::RSDual';
        
        my %nargs = (
                     GZERR       => $self->{GZERR},
                     rs          => $rsd_tv,
                     select_list => \@sel_list,
                     # NOTE: alias is now a required argument for
                     # RSExpr, even though the DUAL rowsource cannot
                     # have name column expressions.
                     alias       => $tabinfo->{tc_table_fullname}
                     );
        my %rsx_h;
        my $rsx_tv = tie %rsx_h, 'Genezzo::Row::RSExpr', %nargs;
        
        my $sth = $rsx_tv->SQLPrepare();
        
        return ("vanilla", $sth);

    }
    elsif (ref($alg->{sql_insert}->[1]->{insert_values}) eq 'HASH')
    {
        my %q1 = (
                  orderby_clause => [],
                  sql_query      =>  $alg->{sql_insert}->[1]->{insert_values}
                  );
        return ("insert select", \%q1);
    }

    $msg = "cannot execute INSERT command";
    %earg = (self => $self, msg => $msg,
             severity => 'warn');
    
    &$GZERR(%earg)
        if (defined($GZERR));
    
    return undef;
} # end SQLInsert

END { }       # module clean-up code here (global destructor)

## YOUR CODE GOES HERE

1;  # don't forget to return a true value from the file

__END__
# Below is stub documentation for your module. You better edit it!

=head1 NAME

Genezzo::XEval - Execution and Expression Evaluation

=head1 SYNOPSIS

use Genezzo::XEval;


=head1 DESCRIPTION

Perform expression evaluation and command execution.

=head1 ARGUMENTS

=head1 FUNCTIONS

=over 4

=item Dict

get or set the dictionary object

=item SQLAlter

entry point for SQL Alter commands, e.g. ALTER TABLE 

=item SQLInsert

Execute SQL INSERT

=back

=head2 EXPORT

=over 4


=back


=head1 LIMITATIONS


=head1 TODO

=over 4

=item Should become more of a dispatch routine, with major guts for
each function stashed in separate modules under XEval.

=item SQLAlter, SQLInsert: move type checking to TypeCheck module.

=back

=head1 AUTHOR

Jeffrey I. Cohen, jcohen@genezzo.com

=head1 SEE ALSO

L<perl(1)>.

Copyright (c) 2005 Jeffrey I Cohen.  All rights reserved.

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA

Address bug reports and comments to: jcohen@genezzo.com

For more information, please visit the Genezzo homepage 
at L<http://www.genezzo.com>

=cut
